flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3532,11 +3573,11 @@ class ArrayExprLowering {
           // Store the element in the buffer.
           auto buff =
               builder.createConvert(loc, fir::HeapType::get(resTy), mem);
-          auto buffi = builder.create<fir::CoordinateOp>(loc, eleRefTy, buff,
-                                                         mlir::ValueRange{off});
-          auto castedStr = builder.createConvert(loc, eleRefTy, v.getAddr());
-          auto load = builder.create<fir::LoadOp>(loc, castedStr);
-          builder.create<fir::StoreOp>(loc, load, buffi);
+          auto buffi = computeCoordinate(buff, off);
+          auto args = Fortran::lower::createArguments(
+              builder, loc, memcpyType(), buffi, v.getAddr(), eleSz,
+              /*volatile=*/builder.createBool(loc, false));
+          createCallMemcpy(args);
```

Comment:
I think for derived types, doing a memcpy is not always the right thing. If they are allocated allocatable components, then a deep copy will be required, This deep copy can of course lead to more deep copies if the allocatable itself is a derived type with allocatable components and so on.

I am also not sure user defined assignment shouldn't be considered when making such copies (I did not find any clear requirements yet.).
Steve and Peter's proposal is worth considering for such derived types (that do not have to be derived types with length parameters). Given the design choice for automatic components in PDT (making them allocatable components), the same will apply.

Reply:
I'll add a TODO for the allocatable case. This doesn't attempt to implement that.